### PR TITLE
feat: insert database row into selected database

### PR DESCRIPTION
**Pull request by speed2exe**
## API for Row Insertion
- Allow user to do a HTTP request to insert a row into a given database (`workspace_id`, `database_id`)

### Task
- [x] Server Implementation
- [x] Client API
- [x] Test cases

Requires this PR to be merged first: https://github.com/AppFlowy-IO/AppFlowy-Collab/pull/344

### CODE REVIEW

File: src/biz/collab/ops.rs [hunk 1]
```diff
@@ -498,6 +476,142 @@ pub async fn list_database_row_ids(
...
 
+pub async fn insert_database_row(
+  collab_storage: &CollabAccessControlStorage,
+  pg_pool: &PgPool,
+  workspace_uuid_str: &str,
+  database_uuid_str: &str,
+  uid: i64,
+  cell_value_by_id: HashMap<String, serde_json::Value>,
+) -> Result<String, AppError> {
+  let mut db_txn = pg_pool.begin().await?;
```
Comment (appflowy) [inline src/biz/collab/ops.rs:487]: can move to line 547?

File: src/biz/collab/ops.rs [hunk 2]
```diff
@@ -498,6 +476,142 @@ pub async fn list_database_row_ids(
...
+
+  let all_fields = db_body.fields.get_all_fields(&db_collab.transact());
+  let field_by_id = all_fields.iter().fold(HashMap::new(), |mut acc, field| {
+    acc.insert(field.id.clone(), field.clone());
+    acc
+  });
+  let type_option_reader_by_id = type_option_writer_by_id(&all_fields);
+  let field_by_name = field_by_name_uniq(all_fields);
+
+  let new_db_row_id = uuid::Uuid::new_v4().to_string();
```
Comment (appflowy) [inline src/biz/collab/ops.rs:501]: we might have a function call: gen_row_id()

File: src/biz/collab/ops.rs [hunk 3]
```diff
@@ -521,6 +635,86 @@ pub async fn get_database_fields(
...
+  collab_storage: &CollabAccessControlStorage,
+  pg_pool: &PgPool,
+  workspace_id: &str,
+  database_id: &str,
+  insert_field: AFInsertDatabaseField,
+) -> Result<String, AppError> {
+  let (mut db_collab, db_body) =
+    get_database_body(collab_storage, workspace_id, database_id).await?;
+
+  let new_id = uuid::Uuid::new_v4()
```
Comment (appflowy) [inline src/biz/collab/ops.rs:651]: why we need to do this? You can use nanoid id or just call `gen_field_id`

File: src/biz/collab/ops.rs [hunk 4]
```diff
@@ -521,6 +635,86 @@ pub async fn get_database_fields(
...
+  let new_field = Field {
+    id: new_id.clone(),
+    name: insert_field.name,
+    field_type: insert_field.field_type,
+    type_options,
+    ..Default::default()
+  };
+
+  let db_collab_update = {
+    let mut yrs_txn = db_collab.transact_mut();
```
Comment (appflowy) [inline src/biz/collab/ops.rs:680]: lots of things you do it manually were already implement. For example create a field and update its views.
image


If we can't use Database directly, maybe we can do some refactor in collab-repo.

File: src/biz/collab/utils.rs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+      _ => {
+        result.insert(key.clone(), serde_json::to_value(value).unwrap_or_default());
+      },
+    }
+  }
+
+  result
+}
+
+pub fn collab_from_doc_state(doc_state: Vec<u8>, object_id: &str) -> Result<Collab, AppError> {
```
Comment (appflowy) [inline src/biz/collab/utils.rs:179]: Duplicated code. Please check biz::collab::ops

File: src/biz/collab/ops.rs [hunk 6]
```diff
@@ -521,6 +635,86 @@ pub async fn get_database_fields(
...
+    }
+    db_body.views.clear(&mut yrs_txn);
+    for view in db_views {
+      db_body.views.insert_view(&mut yrs_txn, view);
+    }
+
+    yrs_txn.encode_update_v1()
+  };
+
+  let updated_db_collab = encode_collab_v1_bytes(&db_collab, CollabType::Database)?;
```
Comment (appflowy) [inline src/biz/collab/ops.rs:694]: please check `collab_to_bin`

Comment (speed2exe) [reply src/biz/collab/ops.rs:694]: That method is private, I'll move to utils and make it public.

File: src/biz/collab/utils.rs (hunk 5 above)
Comment (speed2exe) [reply src/biz/collab/utils.rs:179]: I simply move this function from there to `utils.rs`. it's not duplicated

File: src/biz/collab/ops.rs (hunk 1 above)
Comment (speed2exe) [reply src/biz/collab/ops.rs:487]: I'll move it down to about 543, as this op also requires inserting row collab, which should use the same transaction

File: src/biz/collab/ops.rs (hunk 4 above)
Comment (speed2exe) [reply src/biz/collab/ops.rs:675]: I've used this

File: src/biz/collab/ops.rs (hunk 3 above)
Comment (speed2exe) [reply src/biz/collab/ops.rs]: I've used this

File: src/biz/collab/ops.rs (hunk 2 above)
Comment (speed2exe) [reply src/biz/collab/ops.rs]: I've used this

File: src/biz/collab/ops.rs (hunk 1 above)
Comment (speed2exe) [reply src/biz/collab/ops.rs]: Some additional modification: I have move all the collab non db ops to the top, which further shorten the txn duration.

File: src/biz/collab/utils.rs [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+  .ok_or_else(|| {
+    AppError::Internal(anyhow::anyhow!(
+      "Failed to create database body from collab, db_collab_id: {}",
+      database_uuid_str,
+    ))
+  })?;
+  Ok((db_collab, db_body))
+}
+
+pub fn encode_collab_v1_bytes(
```
Comment (appflowy) [inline src/biz/collab/utils.rs:206]: what is the difference between collab_to_bin and encode_collab_v1_bytes ?

File: src/biz/collab/utils.rs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+  type_option_reader_by_id: &HashMap<String, Box<dyn TypeOptionCellReader>>,
+) -> HashMap<String, HashMap<String, serde_json::Value>> {
+  let mut row_details: HashMap<String, HashMap<String, serde_json::Value>> =
+    HashMap::with_capacity(cells.len());
+
+  for (field_id, field) in field_by_id_name_uniq {
+    let cell: Cell = match cells.get(field_id) {
+      Some(cell) => cell.clone(),
+      None => {
+        tracing::error!("Failed to get cell by field id: {}", field.id);
```
Comment (appflowy) [inline src/biz/collab/utils.rs:55]: it might be empty if there are no values in the cell. So it's not a error.

File: src/biz/collab/utils.rs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+pub fn type_option_writer_by_id(
+  fields: &[Field],
+) -> HashMap<String, Box<dyn TypeOptionCellWriter>> {
+  let mut type_option_reader_by_id: HashMap<String, Box<dyn TypeOptionCellWriter>> =
+    HashMap::with_capacity(fields.len());
+  for field in fields {
+    let field_id: String = field.id.clone();
+    let type_option_reader: Box<dyn TypeOptionCellWriter> = {
+      let field_type: &FieldType = &FieldType::from(field.field_type);
+      let type_option_data: TypeOptionData = match field.type_options.get(&field_type.type_id()) {
```
Comment (appflowy) [inline src/biz/collab/utils.rs:119]: there is a method call get `get_any_type_option` in `Field`

File: src/biz/collab/utils.rs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+pub fn type_option_reader_by_id(
+  fields: &[Field],
+) -> HashMap<String, Box<dyn TypeOptionCellReader>> {
+  let mut type_option_reader_by_id: HashMap<String, Box<dyn TypeOptionCellReader>> =
+    HashMap::with_capacity(fields.len());
+  for field in fields {
+    let field_id: String = field.id.clone();
+    let type_option_reader: Box<dyn TypeOptionCellReader> = {
+      let field_type: &FieldType = &FieldType::from(field.field_type);
+      let type_option_data: TypeOptionData = match field.type_options.get(&field_type.type_id()) {
```
Comment (appflowy) [inline src/biz/collab/utils.rs:140]: Same as above

File: src/biz/collab/utils.rs (hunk 7 above)
Comment (speed2exe) [reply src/biz/collab/utils.rs:206]: `encode_collab_v1_bytes` spawns a tokio task blocking task to do it, while `encode_collab_v1_bytes` dont. I am removing this function, as it's no longer being used, (have switched to use `collab_to_bin` instead).

**Pull request merged**